tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -5822,6 +5832,560 @@ Status ConvertBatchMatMul(OpConverterParams* params) {
+  if ((*operand)->is_weights()) {
+    nvinfer1::Dims new_dims;
+    TF_RETURN_IF_ERROR(GetEinsumNewStaticShape(*desc, &new_dims));
+    if (!need_transpose) {
+      TRT_ShapedWeights weights((*operand)->weights());
+      TF_RETURN_IF_ERROR(weights.SetShape(new_dims));
+      operand->reset(new TRT_TensorOrWeights(weights));
+      return Status::OK();
+    }
+    // TODO(tfeher): Instead of creating a tensor that will be transposed,
+    // transpose the weight itself. Keeping it weight could enable FC layer.
+    nvinfer1::ITensor* tensor = params->converter->CreateConstantLayer(
+        (*operand)->weights(), (*operand)->GetTrtDims());
+    operand->reset(new TRT_TensorOrWeights(tensor));
+  }
+
+  if (!need_transpose && !need_reshape) return Status::OK();
+  nvinfer1::ITensor* operand_tensor = (*operand)->tensor();
+  TFTRT_RETURN_ERROR_IF_NULLPTR(operand_tensor, "Null tensor at Einsum");
+  nvinfer1::IShuffleLayer* layer =
+      params->converter->network()->addShuffle(*operand_tensor);
+
+  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, params->node_def.name());
+  params->converter->SetLayerName(layer, params->node_def, "shuffle",
+                                  /*op_instance=*/op_instance);
+  // Set new shape.
+  if (need_reshape) {
+    if (desc->HasStaticShape()) {
+      nvinfer1::Dims new_dims;
+      TF_RETURN_IF_ERROR(GetEinsumNewStaticShape(*desc, &new_dims));
+      layer->setReshapeDimensions(new_dims);
+    } else {
+      nvinfer1::ITensor* new_shape;
+      TF_RETURN_IF_ERROR(GetEinsumNewDynamicShape(params, *desc, &new_shape));
+      layer->setInput(1, *new_shape);
+    }
+  }
+
+  if (need_transpose) {
+    layer->setFirstTranspose(desc->GetPermutation());
+  }
+  operand->reset(new TRT_TensorOrWeights(layer->getOutput(0)));
+  return Status::OK();
+}
+
+// Combines output dims/labels by coping batch and free dims/labels from input
+// A, and concatenateing free values from input B.
+template <typename InputIterator, typename OutputIterator>
+void AssembleOutput(InputIterator begin_a, InputIterator begin_b,
+                    const EinsumDescriptor& desc_a,
+                    const EinsumDescriptor& desc_b, OutputIterator out) {
+  std::copy(begin_a, begin_a + desc_a.b, out);
+  begin_a += desc_a.offset_f;
+  std::copy(begin_a, begin_a + desc_a.f, out + desc_a.b);
+  begin_b += desc_b.offset_f;
+  std::copy(begin_b, begin_b + desc_b.f, out + desc_a.b + desc_a.f);
+}
+
+Status ShuffleEinsumOutput(OpConverterParams* params, EinsumDescriptor desc_a,
+                           EinsumDescriptor desc_b,
+                           const std::vector<int>& permutation,
+                           nvinfer1::ITensor** output) {
+  // The output has the following dimensions: NAB where N represents the batch
+  // dimensions (0 or more), A and B are a single output dimensions, they are
+  // equal to the product of free dimensions of inputs a and b. First we
+  // expand the free dimensions.
```

Comment:
Let's revise this comment and move it outside the routine to describe what this routine does. I only has the beginning part below:
The raw output dimension from MatMul is [B,X], where B is the batch dimension and X=F_a*F_b is the free dimension and equal to the product of the free dimensions in a and b.

Reply:
There are two free dims, one from each input. I have updated the docstring to explain this.